### PR TITLE
Filter out info from non-const initializers during shape inferencing 

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
* Don't return shape for non-const initializer in InferenceContextImpl::getInputType
* Don't return non-const initializer in InferenceContextImpl::getInputData
* Update graph_utils to support these scenarios
   - fix GetConstantInitializer to make sure a name is for an outer scope value before checking a parent graph, as a local name could shadow an outer scope initializer.
* Add unit test

**Motivation and Context**
We potentially invalidate allocation planning and buffer re-use if during shape inferencing we use info from a non-const initializer. Such an initializer could get overridden at runtime by input of a different shape.

### CODE REVIEW

File: onnxruntime/core/graph/graph.cc [hunk 1]
```diff
@@ -1276,15 +1278,23 @@ class InferenceContextImpl : public ONNX_NAMESPACE::InferenceContext {
...
+
+      // if the input is an initializer that is not const we have to strip out the shape as an
+      // override via a graph input may change it
+      if (type != nullptr && graph_utils::IsInitializer(graph_, p_node_arg->Name(), true) &&
+          !graph_utils::IsConstantInitializer(graph_, p_node_arg->Name(), true)) {
+        TypeProto temp_type(*type);
+        temp_type.mutable_tensor_type()->clear_shape();
+        temp_types_.push_back(std::move(temp_type));
+        type = &temp_types_.back();
+      }
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxruntime/core/graph/graph.cc]: I think we should probably delete this line too: https://github.com/microsoft/onnxruntime/blob/166b1f86db0d9a41189a65b1bfb30110424cc955/onnxruntime/core/graph/graph.cc#L1635 .

[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxruntime/core/graph/graph.cc]: When we create a NodeArg for a graph input, however, it seems appropriate to update the shape (if the graph input has it). The idea with the shapes specified for graph-inputs is that they are constraints on the real-inputs supplied at inference-time: the planner/optimizer are free to use this information for optimization purposes, and at inference-time if a mismatching input is supplied, we should check and report an error. Given that this is what will (or should) happen for non-initializer graph-inputs, it seems the most appropriate treatment for initializer graph-inputs too, right?

[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/graph/graph.cc]: Updating to process graph inputs first so the graph input's shape is preferred if there's a matching initializer. 
Graph::TypeCheckInputsAndInitializers will error out if there's a mismatch between shapes.
Don't need to remove the shape from what's returned by the inferencing context with these changes.

File: onnxruntime/core/graph/graph.cc [hunk 2]
```diff
@@ -1631,12 +1635,17 @@ common::Status Graph::TypeCheckInputsAndInitializers() {
...
       }
       const TensorShapeProto* p_existing_shape = node_arg->Shape();
-      if (nullptr == p_existing_shape)
-        node_arg->SetShape(inferred_shape);
-      else {
+      if (nullptr == p_existing_shape) {
+        // use the inferred shape if this is a constant initializer (cannot be overridden).
+        // if not it has a matching graph input, and we prefer the shape info (or lack of info) from the graph input
+        if (graph_utils::IsConstantInitializer(*this, name, false)) {
+          node_arg->SetShape(inferred_shape);
```
Comment (gramalingam) [inline onnxruntime/core/graph/graph.cc]: The logic looks fine. However, I wonder now whether there is some redundancy between this and the earlier logic when the NodeArgs are created (around new line 688)? Do we need to do this in both places, or is one of this redundant?

Comment (skottmckay) [reply onnxruntime/core/graph/graph.cc]: As discussed, there's still scenarios where shape could be nullptr at this point and needs to be set.

**Pull request merged**